torch/utils/mem_tracker.py

```diff
@@ -0,0 +1,510 @@
+         and each value is another dictionary with keys as memory reference types (e.g., parameters, buffers,
+         activations, gradients, optimizer states) and values as the memory consumed in bytes.
+    """
+
+    def __init__(self, mod_fqn: str):
+        self.mod_fqn = mod_fqn
+        self.parameter_mem: int
+        self.buffer_mem: int
+        self.input_mem: int
+        self.output_mem: int
+        self.local_max: Dict[torch.device, int]
+        self.snapshots: Dict[ModState, Dict[torch.device, Dict[str, int]]] = {}
+
+
+@dataclass
+class WeakRefInfo:
+    def __init__(
+        self, size: int, element_size: int, device: torch.device, reftype: RefType
+    ) -> None:
+        self.size = size
+        self.element_size = element_size
+        self.reftype = reftype
+        self.device = device
+        self.mem_consumed = self._calculate_mem_consumed()
+
+    def _calculate_mem_consumed(self) -> int:
+        mem = self.size * self.element_size
+        if self.device.type == "cuda":
+            return math.ceil((mem) / _PYTORCH_MIN_ALLOCATE) * _PYTORCH_MIN_ALLOCATE
+        return mem
+
+    def get_mem_consumed(self, st: torch.UntypedStorage) -> int:
+        if st.size() != self.size:
+            self.size = st.size()
+            self.mem_consumed = self._calculate_mem_consumed()
+        return self.mem_consumed
+
+    @staticmethod
+    def get_untyped_storage(t: torch.Tensor) -> torch.UntypedStorage:
+        if isinstance(t, DTensor):
+            t = t._local_tensor
+        return t.untyped_storage()
+
+    @classmethod
+    def get_winfo_if_exists(
+        cls, st: torch.UntypedStorage, WINFO: WeakIdKeyDictionary
+    ) -> Optional[Self]:
+        winfo = WINFO.get(st, None)
+        if winfo is None:
+            return None
+        return winfo
+
+    @classmethod
+    def create_winfo(
+        cls, st: torch.UntypedStorage, reftype: RefType, WINFO: WeakIdKeyDictionary
+    ) -> Self:
+        winfo = cls(st.size(), st.element_size(), st.device, reftype)
+        WINFO[st] = winfo
+        return winfo
+
+    @classmethod
+    def update_and_maybe_create_winfo(
+        cls,
+        t: torch.Tensor,
+        reftype: RefType,
+        WINFO: WeakIdKeyDictionary,
+        update_existing: bool = False,
+    ) -> Self:
+        st = cls.get_untyped_storage(t)
+        if winfo := cls.get_winfo_if_exists(st, WINFO):
+            winfo.reftype = reftype
+            return winfo
+        elif update_existing:
+            raise ValueError("No existing winfo found")
+        else:
+            winfo = cls.create_winfo(st, reftype, WINFO)
+            return winfo
+
+
+def _get_mem_divisor(units: str) -> int:
+    return {"B": 1, "KB": 2**10, "MB": 2**20, "GB": 2**30}.get(units, 1)
+
+
+def _rounding_fn(value: int, divisor: int, precision: int) -> Union[float, int]:
+    return value if divisor == 1 else round(value / divisor, precision)
```

Comment:
not sure if we really need to move out of line these one-line functions?

Reply:
I kept them for re-use in fsdp mem tracker, otherwise I would have to duplicate this code.

Reply:
So the fsdp version will depend on private API within this file?

Reply:
So what do you suggest? Making these methods public? or just replicating them in FSDP version?